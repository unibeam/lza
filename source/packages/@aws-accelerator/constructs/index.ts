/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

export * from './lib/aws-accelerator/get-accelerator-metadata';
export * from './lib/aws-budgets/budget-definition';
export * from './lib/aws-cloudformation/get-resource-type';
export * from './lib/aws-cloudwatch-logs/cloudwatch-log-data-protection';
export * from './lib/aws-cloudwatch-logs/cloudwatch-destination';
export * from './lib/aws-cloudwatch-logs/cloudwatch-log-group';
export * from './lib/aws-cloudwatch-logs/cloudwatch-logs-subscription-filter';
export * from './lib/aws-configservice/config-tags';
export * from './lib/aws-configservice/config-aggregation';
export * from './lib/aws-configservice/config-recorder';
export * from './lib/data-perimeter/detect-resource-policy';
export * from './lib/data-perimeter/remediate-resource-policy';
export * from './lib/data-perimeter/remediation-ssm-document';
export * from './lib/aws-controltower/create-accounts';
export * from './lib/aws-controltower/create-enabled-controls';
export * from './lib/aws-cur/report-definition';
export * from './lib/aws-detective/detective-graph-config';
export * from './lib/aws-detective/detective-members';
export * from './lib/aws-detective/detective-organization-admin-account';
export * from './lib/aws-directconnect/direct-connect-gateway';
export * from './lib/aws-directconnect/gateway-association';
export * from './lib/aws-directconnect/virtual-interface';
export * from './lib/aws-directory-service/active-directory';
export * from './lib/aws-directory-service/active-directory-configuration';
export * from './lib/aws-directory-service/active-directory-log-subscription';
export * from './lib/aws-directory-service/active-directory-resolver-rule';
export * from './lib/aws-directory-service/share-active-directory';
export * from './lib/aws-ec2/account-warming';
export * from './lib/aws-ec2/cross-account-route';
export * from './lib/aws-ec2/customer-gateway';
export * from './lib/aws-ec2/delete-default-vpc';
export * from './lib/aws-ec2/dhcp-options';
export * from './lib/aws-ec2/ebs-encryption';
export * from './lib/aws-ec2/firewall';
export * from './lib/aws-ec2/firewall-config-replacements';
export * from './lib/aws-ec2/firewall-instance';
export * from './lib/aws-ec2/firewall-asg';
export * from './lib/aws-ec2/ipam';
export * from './lib/aws-ec2/ipam-pool';
export * from './lib/aws-ec2/ipam-organization-admin-account';
export * from './lib/aws-ec2/ipam-scope';
export * from './lib/aws-ec2/ipam-subnet';
export * from './lib/aws-ec2/prefix-list';
export * from './lib/aws-ec2/prefix-list-route';
export * from './lib/aws-ec2/route-table';
export * from './lib/aws-ec2/subnet-id-lookup';
export * from './lib/aws-ec2/transit-gateway';
export * from './lib/aws-ec2/transit-gateway-connect';
export * from './lib/aws-ec2/transit-gateway-peering';
export * from './lib/aws-ec2/transit-gateway-prefix-list-reference';
export * from './lib/aws-ec2/transit-gateway-route-table';
export * from './lib/aws-ec2/transit-gateway-static-route';
export * from './lib/aws-ec2/vpc';
export * from './lib/aws-ec2/vpc-endpoint';
export * from './lib/aws-ec2/vpc-id-lookup';
export * from './lib/aws-ec2/vpc-peering';
export * from './lib/aws-ec2/vpn-connection';
export * from './lib/aws-events/move-account-rule';
export * from './lib/aws-events/new-cloudwatch-log-event-rule';
export * from './lib/aws-events/revert-scp-changes';
export * from './lib/aws-elasticloadbalancingv2/gateway-load-balancer';
export * from './lib/aws-elasticloadbalancingv2/nlb-addresses';
export * from './lib/aws-guardduty/guardduty-detector-config';
export * from './lib/aws-guardduty/guardduty-members';
export * from './lib/aws-guardduty/guardduty-organization-admin-account';
export * from './lib/aws-guardduty/guardduty-publishing-destination';
export * from './lib/aws-auditmanager/auditmanager-organization-admin-account';
export * from './lib/aws-auditmanager/auditmanager-reports-destination';
export * from './lib/aws-detective/detective-members';
export * from './lib/aws-detective/detective-organization-admin-account';
export * from './lib/aws-detective/detective-graph-config';
export * from './lib/aws-iam/password-policy';
export * from './lib/aws-iam/service-linked-role';
export * from './lib/aws-identity-center/identity-center-instance';
export * from './lib/aws-identity-center/identity-center-organization-admin-account';
export * from './lib/aws-identity-center/identity-center-get-permission-set-role-arn';
export * from './lib/aws-identity-center/identity-center-assignments';
export * from './lib/aws-firehose/cloudwatch-to-s3-firehose';
export * from './lib/aws-fms/fms-notification-channel';
export * from './lib/aws-fms/fms-organization-admin-account';
export * from './lib/aws-kms/key-encryption';
export * from './lib/aws-kms/key-lookup';
export * from './lib/aws-macie/macie-export-config-classification';
export * from './lib/aws-macie/macie-members';
export * from './lib/aws-macie/macie-organization-admin-account';
export * from './lib/aws-macie/macie-session';
export * from './lib/aws-networkfirewall/firewall';
export * from './lib/aws-networkfirewall/policy';
export * from './lib/aws-networkfirewall/rule-group';
export * from './lib/aws-opt-in-regions/enable-opt-in-regions';
export * from './lib/aws-organizations/account';
export * from './lib/aws-organizations/create-accounts';
export * from './lib/aws-organizations/enable-aws-service-access';
export * from './lib/aws-organizations/enable-policy-type';
export * from './lib/aws-organizations/move-accounts';
export * from './lib/aws-organizations/organizational-units';
export * from './lib/aws-organizations/policy';
export * from './lib/aws-organizations/policy-attachment';
export * from './lib/aws-organizations/register-delegated-administrator';
export * from './lib/aws-organizations/validate-scp-count';
export * from './lib/aws-ram/enable-sharing-with-aws-organization';
export * from './lib/aws-ram/resource-share';
export * from './lib/aws-ram/share-subnet-tags';
export * from './lib/aws-route-53-resolver/firewall-domain-list';
export * from './lib/aws-route-53-resolver/firewall-rule-group';
export * from './lib/aws-route-53-resolver/query-logging-config';
export * from './lib/aws-route-53-resolver/resolver-endpoint';
export * from './lib/aws-route-53-resolver/resolver-rule';
export * from './lib/aws-route-53/associate-hosted-zones';
export * from './lib/aws-route-53/hosted-zone';
export * from './lib/aws-route-53/record-set';
export * from './lib/aws-s3/bucket';
export * from './lib/aws-s3/bucket-encryption';
export * from './lib/aws-s3/bucket-replication';
export * from './lib/aws-s3/bucket-policy';
export * from './lib/aws-s3/bucket-prefix';
export * from './lib/aws-s3/central-logs-bucket';
export * from './lib/aws-s3/public-access-block';
export * from './lib/aws-s3/validate-bucket';
export * from './lib/aws-securityhub/securityhub-members';
export * from './lib/aws-securityhub/securityhub-organization-admin-account';
export * from './lib/aws-securityhub/securityhub-standards';
export * from './lib/aws-securityhub/securityhub-region-aggregation';
export * from './lib/aws-servicecatalog/get-portfolio-id';
export * from './lib/aws-service-quota/limits-service-quota-definition';
export * from './lib/aws-servicecatalog/share-portfolio-with-org';
export * from './lib/aws-servicecatalog/propagate-portfolio-associations';
export * from './lib/aws-ssm/document';
export * from './lib/aws-ssm/inventory';
export * from './lib/aws-ssm/session-manager-settings';
export * from './lib/aws-ssm/policy-attachment';
export * from './lib/aws-ssm/put-ssm-parameter';
export * from './lib/aws-ssm/ssm-parameter-lookup';
export * from './lib/aws-events/security-hub-events-log';
export * from './lib/aws-ec2/create-launch-template';
export * from './lib/aws-autoscaling/create-autoscaling-group';
export * from './lib/aws-elasticloadbalancingv2/target-group';
export * from './lib/aws-elasticloadbalancingv2/network-load-balancer';
export * from './lib/aws-elasticloadbalancingv2/application-load-balancer';
export * from './lib/aws-certificate-manager/certificate';
export * from './lib/lza-custom-resource';
export * from './lib/lza-lambda';
