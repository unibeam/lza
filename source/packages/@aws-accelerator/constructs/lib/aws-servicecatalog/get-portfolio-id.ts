/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { v4 as uuidv4 } from 'uuid';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Get the PortfolioId from servicecatalog
 */
export interface GetPortfolioIdProps {
  readonly displayName: string;
  readonly providerName: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

export class GetPortfolioId extends Construct {
  public readonly portfolioId: string;

  constructor(scope: Construct, id: string, props: GetPortfolioIdProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::GetPortfolioId';

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'get-portfolio-id/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Sid: 'ServiceCatalog',
          Effect: 'Allow',
          Action: ['servicecatalog:ListPortfolios'],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        displayName: props.displayName,
        providerName: props.providerName,
        uuid: uuidv4(), // Generates a new UUID to force the resource to update
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.portfolioId = resource.ref;
  }
}
