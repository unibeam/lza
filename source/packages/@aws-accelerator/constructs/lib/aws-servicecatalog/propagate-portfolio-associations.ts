/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { PortfolioConfig } from '@aws-accelerator/config';
import { v4 as uuidv4 } from 'uuid';
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * This construct enables the propagation of Service Catalog Portfolio
 */
export interface PropagatePortfolioAssociationsProps {
  /**
   * A list of account Ids the portfolio is shared with
   */
  readonly shareAccountIds: string[];
  /**
   * The name of the role to assume from the account containing the portfolio
   */
  readonly crossAccountRole: string;
  /**
   * Id of the portfolio to be shared
   */
  readonly portfolioId: string;
  /**
   * The portfolio definition from the customizations-config.yaml file
   */
  readonly portfolioDefinition: PortfolioConfig;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

export class PropagatePortfolioAssociations extends Construct {
  constructor(scope: Construct, id: string, props: PropagatePortfolioAssociationsProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::PropagatePortfolioAssociations';

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'propagate-portfolio-associations/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Sid: 'ServiceCatalog',
          Effect: 'Allow',
          Action: ['sts:AssumeRole'],
          Resource: '*',
        },
      ],
    });

    const customResourceObjects = [];
    customResourceObjects.push(
      new cdk.CustomResource(this, `PropagateAssociations`, {
        resourceType: RESOURCE_TYPE,
        serviceToken: provider.serviceToken,
        properties: {
          crossAccountRole: props.crossAccountRole,
          portfolioId: props.portfolioId,
          portfolioDefinition: JSON.stringify(props.portfolioDefinition),
          shareAccountIds: props.shareAccountIds.join(','),
          partition: cdk.Stack.of(this).partition,
          uuid: uuidv4(), // Generates a new UUID to force the resource to update
        },
      }),
    );

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });

    for (const customResource of customResourceObjects) {
      customResource.node.addDependency(logGroup);
    }
  }
}
