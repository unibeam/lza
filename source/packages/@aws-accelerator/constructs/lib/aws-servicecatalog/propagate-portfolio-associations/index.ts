/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * propagate-portfolio-associations - lambda handler
 *
 * @param event
 * @returns
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { getCrossAccountCredentials, setRetryStrategy } from '@aws-accelerator/utils/lib/common-functions';
import { PortfolioAssociationConfig, PortfolioConfig } from '@aws-accelerator/config';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import {
  AcceptPortfolioShareCommand,
  AssociatePrincipalWithPortfolioCommand,
  DisassociatePrincipalFromPortfolioCommand,
  paginateListPrincipalsForPortfolio,
  paginateListAcceptedPortfolioShares,
  ServiceCatalogClient,
} from '@aws-sdk/client-service-catalog';
import { IAMClient, paginateListRoles } from '@aws-sdk/client-iam';
import { AssumeRoleCommandOutput } from '@aws-sdk/client-sts';

const ssoRolePrefix = '/aws-reserved/sso.amazonaws.com/';

export type CrossAccountClient = {
  serviceCatalog: ServiceCatalogClient;
  iam: IAMClient;
};

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  console.log(event);
  const portfolioId = event.ResourceProperties['portfolioId'];
  const propagator = new AssociationPropagator(event);

  try {
    await propagator.processPropagations();
  } catch (error) {
    console.error(`Failed to propagate principal associations for portfolio ${portfolioId}`);
    console.error(error);
    return {
      PhysicalResourceId: 'none',
      Status: 'FAILED',
    };
  }

  return {
    PhysicalResourceId: 'none',
    Status: 'SUCCESS',
  };
}

export class AssociationPropagator {
  private portfolioId: string;
  private crossAccountRole: string;
  private portfolioDefinition: PortfolioConfig;
  private shareAccountIds: string[];
  private region: string;
  private partition: string;
  private associationsToPropagate: PortfolioAssociationConfig[];
  private requestType: string;

  constructor(event: CloudFormationCustomResourceEvent) {
    this.portfolioId = event.ResourceProperties['portfolioId'];
    this.crossAccountRole = event.ResourceProperties['crossAccountRole'];
    this.portfolioDefinition = <PortfolioConfig>JSON.parse(event.ResourceProperties['portfolioDefinition']);
    this.shareAccountIds = event.ResourceProperties['shareAccountIds'].split(',');
    this.region = process.env['AWS_REGION']!;
    this.partition = event.ResourceProperties['partition'];
    this.requestType = event.RequestType;
    this.associationsToPropagate = this.getPrincipalsToPropagate();
  }

  public async processPropagations(): Promise<void> {
    for (const accountId of this.shareAccountIds) {
      console.log(`Assuming role ${this.crossAccountRole} in account ${accountId}`);
      const crossAccountCredential = await getCrossAccountCredentials(
        accountId,
        this.region,
        this.partition,
        this.crossAccountRole,
      );
      const clients = await this.getCrossAccountClients(crossAccountCredential);
      await this.preprocessPortfolio(clients);
      const existingPrincipalArns = await this.getAssociatedPortfolioPrincipals(clients);
      if (existingPrincipalArns.length > 0) {
        console.log(
          `Found the following principals already associated to portfolio ${
            this.portfolioDefinition.name
          }: ${existingPrincipalArns.join(', ')}`,
        );
      }

      switch (this.requestType) {
        case 'Create':
        case 'Update':
          await this.createUpdatePropagations(accountId, clients, existingPrincipalArns);
          break;
        case 'Delete':
          await this.deletePropagations(accountId, clients, existingPrincipalArns);
      }
    }
  }

  private async createUpdatePropagations(
    accountId: string,
    clients: CrossAccountClient,
    existingPrincipalArns: string[],
  ): Promise<void> {
    for (const principalAssociation of this.associationsToPropagate) {
      const principalArn = await this.getPrincipalArnToAssociate(principalAssociation, accountId, clients.iam);
      if (!existingPrincipalArns.includes(principalArn)) {
        console.log(`Associating principal ${principalArn} with portfolio ${this.portfolioDefinition.name}`);
        await this.createPropagatedAssociation(principalArn, clients);
      } else {
        console.log(
          `Principal ${principalAssociation.name} already associated with portfolio ${this.portfolioDefinition.name}, skipping creation.`,
        );
      }
    }
  }

  private async deletePropagations(
    accountId: string,
    clients: CrossAccountClient,
    existingPrincipalArns: string[],
  ): Promise<void> {
    for (const principalAssociation of this.associationsToPropagate) {
      const principalArn = await this.getPrincipalArnToAssociate(principalAssociation, accountId, clients.iam);
      console.warn(
        `deletePropagations existingPrincipalArns: ${JSON.stringify(
          existingPrincipalArns,
        )}, principalArn: ${principalArn}`,
      );
      if (existingPrincipalArns.includes(principalArn)) {
        await this.deletePropagatedAssociation(principalArn, clients);
      } else {
        console.log(
          `Principal ${principalAssociation.name} not associated with portfolio ${this.portfolioDefinition.name}, skipping deletion.`,
        );
      }
    }
  }

  private async preprocessPortfolio(client: CrossAccountClient): Promise<void> {
    const portfolioImported = await this.doesPortfolioExist(this.portfolioId, client);
    if (!portfolioImported) {
      console.log(`Portfolio ${this.portfolioId} not found. Attempting to accept portfolio share.`);
      await this.importPortfolio(client);
    }
  }

  private async importPortfolio(client: CrossAccountClient): Promise<void> {
    try {
      await throttlingBackOff(() =>
        client.serviceCatalog.send(
          new AcceptPortfolioShareCommand({
            PortfolioId: this.portfolioId,
            PortfolioShareType: 'IMPORTED',
          }),
        ),
      );
    } catch (error) {
      console.error(error);
      console.log('Encountered error attempting to accept portfolio share. Continuing.');
    }
  }

  private async doesPortfolioExist(portfolioId: string, client: CrossAccountClient): Promise<boolean> {
    const importedPortfolioIds = await getAcceptedPortfolioIds(client.serviceCatalog);
    if (importedPortfolioIds.includes(portfolioId)) {
      return true;
    } else {
      return false;
    }
  }

  private async createPropagatedAssociation(principalArn: string, client: CrossAccountClient): Promise<void> {
    try {
      await throttlingBackOff(() =>
        client.serviceCatalog.send(
          new AssociatePrincipalWithPortfolioCommand({
            PortfolioId: this.portfolioId,
            PrincipalARN: principalArn,
            PrincipalType: 'IAM',
          }),
        ),
      );
    } catch (error) {
      console.error(error);
      console.log(`Encountered error attempting to associate ${principalArn} with ${this.portfolioId}. Continuing`);
    }
  }

  private async deletePropagatedAssociation(principalArn: string, client: CrossAccountClient): Promise<void> {
    try {
      await throttlingBackOff(() =>
        client.serviceCatalog.send(
          new DisassociatePrincipalFromPortfolioCommand({
            PortfolioId: this.portfolioId,
            PrincipalARN: principalArn,
          }),
        ),
      );
    } catch (error) {
      console.error(error);
      console.log(`Encountered error attempting to disassociate ${principalArn} from ${this.portfolioId}. Continuing`);
    }
  }

  private async getAssociatedPortfolioPrincipals(client: CrossAccountClient): Promise<string[]> {
    const principalList: string[] = [];
    for await (const page of paginateListPrincipalsForPortfolio(
      { client: client.serviceCatalog },
      { PortfolioId: this.portfolioId },
    )) {
      for (const principal of page.Principals ?? []) {
        if (principal.PrincipalType === 'IAM') {
          principalList.push(principal.PrincipalARN!);
        }
      }
    }
    return principalList;
  }

  private getPrincipalsToPropagate(): PortfolioAssociationConfig[] {
    const principalAssociationList = this.portfolioDefinition.portfolioAssociations.filter(
      a => a.propagateAssociation === true,
    );
    return principalAssociationList;
  }

  private async getPrincipalArnToAssociate(
    association: PortfolioAssociationConfig,
    accountId: string,
    iamClient: IAMClient,
  ): Promise<string> {
    const associationType = association.type.toLowerCase();
    let roleArn = '';

    if (associationType === 'permissionset') {
      roleArn = await getPermissionSetRoleArn(association.name, accountId, iamClient);
    } else {
      roleArn = `arn:${this.partition}:iam::${accountId}:${associationType}/${association.name}`;
    }
    return roleArn;
  }

  private async getCrossAccountClients(crossAccountCredentials: AssumeRoleCommandOutput) {
    const credentials = {
      accessKeyId: crossAccountCredentials.Credentials!.AccessKeyId!,
      secretAccessKey: crossAccountCredentials.Credentials!.SecretAccessKey!,
      sessionToken: crossAccountCredentials.Credentials!.SessionToken!,
    };
    const iamClient = new IAMClient({
      credentials,
      retryStrategy: setRetryStrategy(),
    });

    const serviceCatalogClient = new ServiceCatalogClient({
      region: this.region,
      credentials,
      retryStrategy: setRetryStrategy(),
    });
    return {
      iam: iamClient,
      serviceCatalog: serviceCatalogClient,
    };
  }
}

export async function getPermissionSetRoleArn(
  permissionSetName: string,
  account: string,
  iamClient: IAMClient,
): Promise<string> {
  const regex = new RegExp(`AWSReservedSSO_${permissionSetName}_([0-9a-fA-F]{16})`);
  const roles = await getIamRoleList(iamClient, ssoRolePrefix);
  const foundRole = roles.find(role => {
    const match = regex.test(role.RoleName!);
    console.log(`Test ${JSON.stringify(role)} for pattern ${regex} result: ${match}`);
    return role;
  });
  const roleArn = foundRole?.Arn ?? undefined;

  if (roleArn) {
    console.log(`Found provisioned role for permission set ${permissionSetName} with ARN: ${roleArn}`);
  } else {
    throw new Error(`Unable to find provisioned role for permission set ${permissionSetName} in account ${account}`);
  }

  return roleArn;
}

export async function getIamRoleList(iamClient: IAMClient, prefix: string) {
  const roleList = [];
  for await (const page of paginateListRoles({ client: iamClient }, { PathPrefix: prefix })) {
    if (page.Roles) {
      roleList.push(...page.Roles);
    }
  }
  return roleList;
}

export async function getAcceptedPortfolioIds(scClient: ServiceCatalogClient): Promise<string[]> {
  const importedShares = await listAcceptedShares(scClient, 'IMPORTED');
  const organizationShares = await listAcceptedShares(scClient, 'AWS_ORGANIZATIONS');

  return [...importedShares, ...organizationShares];
}

export async function listAcceptedShares(
  scClient: ServiceCatalogClient,
  shareType: 'IMPORTED' | 'AWS_ORGANIZATIONS',
): Promise<string[]> {
  const shareList: string[] = [];
  for await (const page of paginateListAcceptedPortfolioShares(
    { client: scClient },
    { PortfolioShareType: shareType },
  )) {
    if (page.PortfolioDetails) {
      shareList.push(...page.PortfolioDetails.map(a => a.Id!));
    }
  }
  return shareList;
}
