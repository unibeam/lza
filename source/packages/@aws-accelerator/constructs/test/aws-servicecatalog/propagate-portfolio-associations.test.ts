/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { PropagatePortfolioAssociations } from '../../lib/aws-servicecatalog/propagate-portfolio-associations';
import { snapShotTest } from '../snapshot-test';
import { describe } from '@jest/globals';
import { PortfolioConfig } from '@aws-accelerator/config';

const testNamePrefix = 'Construct(PropagatePortfolioAssociations): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();
const portfolioItem = new PortfolioConfig();

new PropagatePortfolioAssociations(stack, 'PropagatePortfolioAssociations', {
  portfolioId: 'portfolioId',
  shareAccountIds: ['222222222222', '333333333333'],
  crossAccountRole: 'AWSAccelerator-CrossAccount-ServiceCatalog-Role',
  portfolioDefinition: portfolioItem,
  logRetentionInDays: 3653,
  kmsKey: new cdk.aws_kms.Key(stack, 'Key', {}),
});

/**
 * PropagatePortfolioAssociations construct test
 */
describe('PropagatePortfolioAssociations', () => {
  snapShotTest(testNamePrefix, stack);
});
