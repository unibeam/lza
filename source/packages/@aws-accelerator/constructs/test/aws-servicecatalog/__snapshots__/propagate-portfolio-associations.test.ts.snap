// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropagatePortfolioAssociations Construct(PropagatePortfolioAssociations):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomPropagatePortfolioAssociationsCustomResourceProviderHandler282C1D38": {
      "DependsOn": [
        "CustomPropagatePortfolioAssociationsCustomResourceProviderRole4C66847F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomPropagatePortfolioAssociationsCustomResourceProviderRole4C66847F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomPropagatePortfolioAssociationsCustomResourceProviderLogGroup44DE568D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "Key961B73FD",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomPropagatePortfolioAssociationsCustomResourceProviderHandler282C1D38",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomPropagatePortfolioAssociationsCustomResourceProviderRole4C66847F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "sts:AssumeRole",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "ServiceCatalog",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Key961B73FD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "PropagatePortfolioAssociationsPropagateAssociationsB8799497": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomPropagatePortfolioAssociationsCustomResourceProviderLogGroup44DE568D",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomPropagatePortfolioAssociationsCustomResourceProviderHandler282C1D38",
            "Arn",
          ],
        },
        "crossAccountRole": "AWSAccelerator-CrossAccount-ServiceCatalog-Role",
        "partition": {
          "Ref": "AWS::Partition",
        },
        "portfolioDefinition": "{"name":"","provider":"","account":"","regions":[],"portfolioAssociations":[],"products":[]}",
        "portfolioId": "portfolioId",
        "shareAccountIds": "222222222222,333333333333",
        "uuid": "REPLACED-UUID",
      },
      "Type": "Custom::PropagatePortfolioAssociations",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
